### ISSUE TITLE
no default __reduce__ due to non-trivial __cinit__ error

### ISSUE THREAD

**Issue by Leesangoh** [kylebd99/SafeBound#1]
Hi, i got this error while running Experiments/BuildExperiments.py

Can you please help me out with this problem?

(TestEnv) /home/SafeBound# python Experiments/BuildExperiments.py
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/opt/conda/envs/TestEnv/lib/python3.10/concurrent/futures/process.py", line 208, in _sendback_result
    result_queue.put(_ResultItem(work_id, result=result,
  File "/opt/conda/envs/TestEnv/lib/python3.10/multiprocessing/queues.py", line 372, in put
    obj = _ForkingPickler.dumps(obj)
  File "/opt/conda/envs/TestEnv/lib/python3.10/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
  File "stringsource", line 2, in pybloomfilter.BloomFilter.__reduce_cython__
TypeError: no default __reduce__ due to non-trivial __cinit__
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/SafeBound/Experiments/BuildExperiments.py", line 37, in <module>
    time, size = build_stats_object(method='SafeBound',
  File "/home/SafeBound/Source/ExperimentUtils/BuildUtils.py", line 459, in build_stats_object
    return build_safe_bound(benchmark, parameters, outputFile)
  File "/home/SafeBound/Source/ExperimentUtils/BuildUtils.py", line 261, in build_safe_bound
    stats = SafeBound(tables, tableNames, joinColumns,
  File "SafeBoundUtils.pyx", line 246, in SafeBoundUtils.SafeBound.__init__
    self.finishOneStatsBuildProcess()
  File "SafeBoundUtils.pyx", line 254, in SafeBoundUtils.SafeBound.finishOneStatsBuildProcess
    self.tableStatsDict[tableName].addFilterColStats(filterCol, statsFuture.result())
  File "/opt/conda/envs/TestEnv/lib/python3.10/concurrent/futures/_base.py", line 439, in result
    return self.__get_result()
  File "/opt/conda/envs/TestEnv/lib/python3.10/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
TypeError: no default __reduce__ due to non-trivial __cinit__

**Comment by kylebd99** [comment]
Hey,

Thanks for trying to use the package! I am on vacation this week, but I
would be happy to help next Monday. I recall there being a bit of
trickiness with that pybloomfilter package, specifically it has trouble
pickling. There’s a version of the package ( I think it’s a public branch)
on GitHub that solves this, but I can’t remember quite what it is right
now. I’ll update the readme when I get back.

Thanks,
Kyle

**Comment by Leesangoh** [comment]
Thank you for reaching out and letting us know about the issue with the pybloomfilter package. I appreciate your willingness to help, and I hope you're enjoying your vacation!!

I'll make a note to check the GitHub repository for the public branch that addresses the pickling problem. In the meantime, I'll keep an eye out for your readme update when you return.

**Comment by kylebd99** [comment]
I just updated the environment.yaml file. Looking back through things, I remembered the issue. In order to use multiprocessing, I had to use the branch of pybloomfiltermmap found here (https://github.com/stevesimmons/pybloomfiltermmap3/tree/pickleable). This is why I included it in the repository directly. However, the conda environment was still pulling the default package from pip which doesn't support pickling (and therefore multiprocessing).  I removed the reference from the conda config, so hopefully it will use the proper version now because it won't download any other versions. 

So, I would recommend removing pybloomfiltermmap3 from your conda environment, then rebuilding it using the setup.py. Let me know if this works!
